src/ndb_p1.c

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
+ * Contact: Roman Gezikov <rgezikov@exafore.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+#include <string.h>
+#include <assert.h>
+#include "ndb.h"
+#include <libswiftnav/constants.h>
+#include "libswiftnav/logging.h"
+#include "timing.h"
+#include "sbp.h"
+#include "sbp_utils.h"
+
+static ephemeris_t ndb_ephemeris[PLATFORM_SIGNAL_COUNT] _CCM;
+static ndb_element_metadata_t ndb_ephemeris_md[PLATFORM_SIGNAL_COUNT] _CCM;
+static almanac_t ndb_almanac[PLATFORM_SIGNAL_COUNT] _CCM;
+static ndb_element_metadata_t ndb_almanac_md[PLATFORM_SIGNAL_COUNT] _CCM;
+#define NDB_EPHE_FILE_NAME   "ephe"
+#define NDB_ALMA_FILE_NAME   "alma"
+
+static ndb_file_t ndb_ephe_file = { .name = NDB_EPHE_FILE_NAME, .fh = -1,
+    .expected_size = sizeof(ndb_file_version)
+        + sizeof(ephemeris_t) * PLATFORM_SIGNAL_COUNT
+        + sizeof(ndb_element_metadata_nv_t) * PLATFORM_SIGNAL_COUNT };
+
+static ndb_file_t ndb_alma_file = { .name = NDB_ALMA_FILE_NAME, .fh = -1,
+    .expected_size = sizeof(ndb_file_version)
+        + sizeof(almanac_t) * PLATFORM_SIGNAL_COUNT
+        + sizeof(ndb_element_metadata_nv_t) * PLATFORM_SIGNAL_COUNT };
+
+typedef struct {
+  ephemeris_t ephe;
+  bool used;
+  ndb_timestamp_t received_at;
+} ephemeris_candidate_t;
+
+typedef enum {
+  EPHE_IDENTICAL,
+  EPHE_NEW_CANDIDATE,
+  EPHE_NEW_TRUSTED,
+  EPHE_CAND_MISMATCH,
+} ndb_ephemeris_status_t;
+
+#define EPHE_CAND_LIST_LEN (MAX_CHANNELS*2/3)
+#define MAX_EPHE_CANDIDATE_AGE 92 /* seconds */
+static ephemeris_candidate_t ephe_candidates[EPHE_CAND_LIST_LEN] _CCM;
+static MUTEX_DECL(cand_list_access);
+
+#define EPHEMERIS_MESSAGE_SPACING_cycle        (200 / NV_WRITE_REQ_TIMEOUT)
+#define EPHEMERIS_TRANSMIT_EPOCH_SPACING_cycle (15000 / NV_WRITE_REQ_TIMEOUT)
+
+enum ndb_op_code ndb_p1_init()
+{
+  u32 i;
+
+  memset(ephe_candidates, 0, sizeof(ephe_candidates));
+
+  memset(ndb_ephemeris, 0, sizeof(ndb_ephemeris));
+  for (i = 0; i < PLATFORM_SIGNAL_COUNT; i++) {
+    ndb_ephemeris[i].sid = sid_from_global_index(i);
+  }
+
+  if (ndb_load_data(&ndb_ephe_file, ndb_ephemeris, ndb_ephemeris_md,
```

Comment:
I was referring to the whole chunk of code: loading, logging the file status, and setting up the metadata. That's going to be the same for all data types, right?